### PR TITLE
Some improvements and unifier fixes

### LINKED ISSUE #1443: Type inference leaves unsolved uvars and crashes
```
module Test
let rec test ps = match ps with
  | [] -> 0
  | a::q ->
    let rec blah i = match i with
      | [] -> 0
      | b::r -> blah r
    in
    if a = 0 then 0
    else test q
```

Failure("(Test.fst(6,16-6,17)) CheckNoUvars: Unexpected unification variable remains: (*?u1736*) _")

Thanks to @R1kM

### PR DESCRIPTION
**Pull request by mtzguido**
Making this branch so @nikswamy can review some of the changes, as they touch core components.

This branch fixes #1443, #1523 and part of #1514 (@aseemr has the other piece for that).

The most sensitive changes, IMO, are the `tc:` ones.

### CODE REVIEW

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 1]
```diff
@@ -561,7 +561,7 @@ and tc_maybe_toplevel_term env (e:term) : term                  (* type-checked
             (* We constrain this unification variable to be of the shape Type u *)
             (* for some new unification variable u *)
             let t, u = U.type_u () in
-            let g_opt = Rel.try_teq true env c'.res_typ t in
+            let g_opt = Rel.get_subtyping_prop env c'.res_typ t in
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs]: I wonder why we don't just set the expected type to type_u in the environment and remove this case analysis

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 2]
```diff
@@ -1694,9 +1694,10 @@ and solve_rigid_flex_or_flex_rigid_subtyping
 
       let tx = UF.new_transaction () in
       begin
-      match solve_t env eq_prob ({wl' with defer_ok=false; attempting=sub_probs}) with
-      | Success _ ->
+      match solve_t env eq_prob ({wl' with defer_ok=false; wl_implicits = []; attempting=sub_probs}) with
+      | Success (_, imps) ->
          let wl = {wl' with attempting=rest} in
+         let wl = {wl with wl_implicits = wl'.wl_implicits @ imps} in
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs]: This looks good.

Although, FYI, the general structure here is similar to what is provided by `try_solve_without_smt_or_else` (also in this module). I wonder if we should refactor a bit.

**Comment by mtzguido** [comment]
Ouch, still slightly broken.. typechecking `examples/native_tactics/Simple.fst` makes #1091 pop up, but only without two phases. I can follow tomorrow, sorry for the noise.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs (hunk 1 above)
Comment (mtzguido) [reply src/typechecker/FStar.TypeChecker.TcTerm.fs]: Sure, but this also looks to be here only to compute an universe level. I think we should just call `env.universe_of`

Comment (mtzguido) [reply src/typechecker/FStar.TypeChecker.TcTerm.fs]: I made that change in f879a0bc3e6d1c3937a309caa3f875dab098ff8a, now in master.

**Comment by mtzguido** [comment]
I think the new failure is more of an instance of #933 than of #1091, i.e. a bad decreases clause, uncovered by the fix to the unifier. I worked around it for now, as it seems caused by a different issue it only happens without two phases, so I don't think it's a big roadblock.

I would merge this, but we could wait until we fix #933/#1091 if you want, Nik.

**Comment by mtzguido** [comment]
I did get a failure in everest when I kicked a CI build, trying to repro locally now. I'll report back here.

**Pull request merged**